qutip/tests/core/data/test_convert.py

```diff
@@ -0,0 +1,71 @@
+import numpy as np
+import pytest
+from scipy import sparse
+from qutip.core.data import to, create, CSR, Dense, dense, csr
+
+
+@pytest.mark.parametrize(['base', 'dtype'], [
+    pytest.param(dense.zeros(2,2), Dense, id='data.Dense'),
+    pytest.param(csr.zeros(2,2), CSR, id='data.CSR'),
+    pytest.param(np.zeros((10, 10), dtype=np.complex128), Dense, id='array'),
+    pytest.param(sparse.eye(10, dtype=np.complex128, format='csr'), CSR,
+                 id='sparse'),
+    pytest.param(np.zeros((10, 10), dtype=np.int32), Dense, id='array'),
+    pytest.param(sparse.eye(10, dtype=np.float, format='csr'), CSR,
+                 id='sparse'),
+])
+def test_create(base, dtype):
+    # The test of exactitude is done in test_csr, test_dense.
+    created = create(base)
+    assert isinstance(created, dtype)
+
+
+@pytest.mark.parametrize(['from_', 'base'], [
+    pytest.param('dense', dense.zeros(2,2), id='from Dense str'),
+    pytest.param('Dense', dense.zeros(2,2), id='from Dense STR'),
+    pytest.param(Dense, dense.zeros(2,2), id='from Dense type'),
+    pytest.param('csr', csr.zeros(2,2), id='from CSR str'),
+    pytest.param('CSR', csr.zeros(2,2), id='from CSR STR'),
+    pytest.param(CSR, csr.zeros(2,2), id='from CSR type'),
+])
+@pytest.mark.parametrize(['to_', 'dtype'], [
+    pytest.param('dense', Dense, id='to Dense str'),
+    pytest.param('Dense', Dense, id='to Dense STR'),
+    pytest.param(Dense, Dense, id='to Dense type'),
+    pytest.param('csr', CSR, id='to CSR str'),
+    pytest.param('CSR', CSR, id='to CSR STR'),
+    pytest.param(CSR, CSR, id='to CSR type'),
+])
+def test_converters(from_, base, to_, dtype):
+    converter = to[to_, from_]
+    assert isinstance(converter(base), dtype)
+    converter = to[to_]
+    assert isinstance(converter(base), dtype)
+    assert isinstance(to(to_, base), dtype)
+
+
+def test_parse():
```

Comment:
@Ericgig I added these tests -- let me know if I left any cases we should check out.

Reply:
We could read type for to.dtype to automatically add new types instead of fixing it to "CRS", "Dense".
I would be one of the only test that test error message... Not sure if I like it.
Testing error is fine but the exact error message?

Reply:
I think all of the `assert to.parse(...) is ...` should be separate parametrised cases of a single test, so they're all guaranteed to run - we get more information on test failures that way, and better reporting from pytest.  As it is, if there's a string casing bug, the exception tests wouldn't run.

We usually (or try to) do something like
```python
@pytest.parametrize(['input', 'type'], [('dense', Dense), ('Dense', Dense), ('DeNsE', Dense), ...])
def test_parse_correct_values(input, type):
    assert data.to.parse(input) is type

@pytest.parametrize(['input', 'exc_class', 'exc_message'], [
    (5, TypeError, "Invalid dtype is neither a type nor a type name: 5"),
    ...
])
def test_parse_raises(input, exc_class, exc_message):
    with pytest.raises(exc_class) as exc:
        data.to.parse(input)
    assert str(exc.value) == exc_message
```